src/singleton_snap.py

```diff
@@ -0,0 +1,293 @@
+            # Exclusive snap install/remove operations here.
+            pass
+
+        # For configuration changes
+        with manager.config_operation("/etc/otelcol/config.yaml"):
+            # Safe config file modifications here.
+            pass
+
+        # For unit tracking
+        manager.register("otelcol")
+        # Use the snap...
+
+        # For unregistering
+        manager.unregister("otelcol")
+
+    Raises:
+        SingletonSnapError on any error.
+    """
+
+    LOCK_FILE_PREFIX = "LCK.."
+    SEPARATOR = "--"
+    LOCK_DIR = "/run/lock/singleton_snaps"
+
+    def __init__(self, unit_name: str):
+        """Initialize the manager with a normalized unit name.
+
+        Args:
+            unit_name: Identifier for the current unit
+        """
+        self.unit_name = self._normalize_name(unit_name)
+        self._ensure_lock_dir_exists()
+
+    @staticmethod
+    def _normalize_name(name: str) -> str:
+        """Normalize names to contain only alphanumerics, _ and -."""
+        return re.sub(r"[^\w-]", "_", name)
+
+    def _ensure_lock_dir_exists(self) -> None:
+        """Ensure the lock directory exists with correct permissions."""
+        try:
+            os.makedirs(self.LOCK_DIR, exist_ok=True)
+            os.chown(self.LOCK_DIR, os.geteuid(), os.getegid())
+        except OSError as e:
+            if e.errno != errno.EEXIST:
+                raise SingletonSnapError(f"Error creating lock directory: {e}") from e
+
+    def _get_registration_file_path(self, snap_name: str) -> str:
+        """Helper method to construct registration file path.
+
+        The registration file name is constructed as:
+        LCK..{snap_name}--{unit_name}
+        """
+        filename = f"{self.LOCK_FILE_PREFIX}{snap_name}{self.SEPARATOR}{self.unit_name}"
+        return os.path.join(self.LOCK_DIR, filename)
+
+    def _parse_unit_name_from_file(self, filename: str, snap_name: str) -> str:
+        """Helper method to extract unit name from registration filename."""
+        prefix = f"{self.LOCK_FILE_PREFIX}{snap_name}{self.SEPARATOR}"
+        return filename[len(prefix) :]
+
+    def register(self, snap_name: str) -> None:
+        """Register current unit as using the specified snap.
+
+        Raises:
+            SingletonSnapError: if there is an I/O related error creating the lock file.
+        """
+        self._update_registration(snap_name, create=True)
+
+    def unregister(self, snap_name: str) -> None:
+        """Unregister current unit from using the specified snap.
+
+        Raises:
+            SingletonSnapError: if there is an I/O related error removing the lock file.
+        """
+        self._update_registration(snap_name, create=False)
+
+    @contextmanager
+    def _lock_directory(self):
+        dir_fd = None
+        try:
+            dir_fd = os.open(self.LOCK_DIR, os.O_RDONLY)
+            fcntl.flock(dir_fd, fcntl.LOCK_SH)
+            yield
+        except TypeError as e:
```

Comment:
It seems odd to me to be catching a `TypeError` (programmer error) and re-raising it as a `SingletonSnapError`. Is it better to just let type errors bubble up?